File: 54.螺旋矩阵.py
# @before-stub-for-debug-begin
# @before-stub-for-debug-end

#
# @lc app=leetcode.cn id=54 lang=python3
#
# [54] 螺旋矩阵
#

# @lc code=start


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m = len(matrix);
        n = len(matrix[0]);

        left = 0
        right = n - 1
        up = 0
        down = m - 1
        res = []
        # idx = 0


        while left <= right and up <= down:

            idx = left
            while idx <= right: 
                res.append(matrix[up][idx])
                idx += 1
            up += 1

            idx = up
            while idx <= down:
                res.append(matrix[idx][right]) 
                idx += 1
            right -= 1

            idx = right
            while idx >= left and up <= down:
                res.append(matrix[down][idx])
                idx -= 1
            down -= 1

            idx = down
            while idx >= up and left <= right:
                res.append(matrix[idx][left])
                idx -= 1
            left += 1

        return res
# @lc code=end

File: 54.螺旋矩阵.c
/*
 * @lc app=leetcode.cn id=54 lang=c
 *
 * [54] 螺旋矩阵
 */

// @lc code=start

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int *spiralOrder(int **matrix, int matrixSize, int *matrixColSize, int *returnSize)
{

    int m = matrixSize;
    int n = *matrixColSize;
    *returnSize = m * n;

    int *res = (int *)malloc(sizeof(int) * (*returnSize));

    int left = 0;
    int right = n - 1;
    int up = 0;
    int down = m - 1;
    int idx = 0;

    while (left <= right && up <= down)
    {
        for (int i = left; i <= right; i++)
            res[idx++] = matrix[up][i];
        up++;

        for (int i = up; i <= down; i++)
            res[idx++] = matrix[i][right];
        right--;

        for (int i = right; i >= left && up <= down; i--)
            res[idx++] = matrix[down][i];
        down--;

        for (int i = down; i >= up && left <= right; i--)
            res[idx++] = matrix[i][left];
        left++;
    }
    return res;
}
// @lc code=end
